### PR TITLE
Add Directory.open('foo')

### PR DESCRIPTION
**Pull request by adamdeprince**

### CODE REVIEW

File: directorymodule.c [hunk 1]
```diff
@@ -193,9 +197,42 @@ entry_directory(EntryObject *self, PyObject *_) {
   strcpy(buffer + PyString_Size(self->path) + 1, &self->dirent.d_name);
   return make_new_directory(NULL, buffer);
 }
-   
 
 PyObject *
+directory_open(DirectoryObject *self, PyObject *py_name) {
+  FILE *fp = NULL;
+  int fd=-1;
```
Comment (srmaxwell3) [inline directorymodule.c]: Formatting.

File: directorymodule.c [hunk 2]
```diff
@@ -193,9 +197,42 @@ entry_directory(EntryObject *self, PyObject *_) {
...
+  
+  if (self->dirp == -1)
+    self->dirp = open(path, O_DIRECTORY | O_RDONLY);
+  if (self->dirp != -1)
+    fd = openat(self->dirp, name, O_CREAT | O_RDWR, 0777);
+  if (fd != -1)
+    fp = fdopen(fd, "r+");
+  Py_END_ALLOW_THREADS
+
+  if (self->dirp == -1 || fd == -1 || !fp)
```
Comment (srmaxwell3) [inline directorymodule.c:224]: So, this test and the code above _require_ that we succeeded in openat(), thereby ignoring the opportunity to proceed with an open() alternative.  Is that intended?

Comment (adamdeprince) [reply directorymodule.c:224]: Why would `open(a + '/' + b)` succeed where `openat(open(a), b)` fails?   If soemthing weird happens, maybe its best to raise an exception?

File: directorymodule.c (hunk 1 above)
Comment (adamdeprince) [reply directorymodule.c]: fixed

**Pull request merged**